jdbc/src/java/org/apache/hive/jdbc/HiveConnection.java

```diff
@@ -896,39 +946,103 @@ private void openSession() throws SQLException {
       openReq.setPassword(sessConfMap.get(JdbcConnectionParams.AUTH_PASSWD));
     }
 
+    //TODO(Vihang): This is a bit hacky. We piggy back on a dummy OpenSession call
+    // to get the redirect response from the server. Instead its probably cleaner to
+    // explicitly do a HTTP post request and get the response.
+    int numRetry = isBrowserAuthMode() ? 2 : 1;
+    for (int i=0; i<numRetry; i++) {
+      try {
+        openSession(openReq);
+      } catch (TException e) {
+        if (isSamlRedirect(e)) {
+          boolean success = doBrowserSSO();
+          if (!success) {
+            String msg = browserClient.getServerResponse() == null
+                || browserClient.getServerResponse().getMsg() == null ? ""
+                : browserClient.getServerResponse().getMsg();
+            throw new SQLException(
+                "Could not establish connection to " + jdbcUriString + ": "
+                    + msg, " 08S01", e);
+          }
+        } else {
+          throw new SQLException(
+              "Could not establish connection to " + jdbcUriString + ": " + e
+                  .getMessage(), " 08S01", e);
+        }
+      }
+    }
+    isClosed = false;
+  }
+
+  private boolean doBrowserSSO() throws SQLException {
     try {
-      TOpenSessionResp openResp = client.OpenSession(openReq);
+      Preconditions.checkNotNull(browserClient);
+      try (IJdbcBrowserClient bc = browserClient) {
+        browserClient.doBrowserSSO();
+        HiveJdbcBrowserServerResponse response = browserClient.getServerResponse();
+        if (response != null) {
+          return response.isSuccessful();
+        }
+        return false;
+      }
+    } catch (Exception ex) {
+      throw new SQLException("Browser based SSO failed: " + ex.getMessage(),
+          " 08S01",
```

Comment:
is this a known error code for failed auth?

Reply:
I could find where to look for these standard error codes. The closest hit online was the https://docs.microsoft.com/en-us/sql/odbc/reference/appendixes/appendix-a-odbc-error-codes?view=sql-server-ver15 which lists 08S01 as "communications like failure". I didn't see anything which suggests a authentication error. The closest was perhaps 28000 but I am not sure if this is the right place to look for these codes. Do you know where can I find this for JDBC?